File: frontend/src/types/activity.ts
export type ActivityCategory =
  | "sightseeing"
  | "food"
  | "adventure"
  | "shopping"
  | "wellness"
  | "cultural"
  | "other";

export interface Activity {
  id: number;
  city_id: number;
  name: string;
  description: string | null;
  category: ActivityCategory;
  cost_usd: number;
  duration_hours: number | null;
  map_link: string | null;
  opening_time: string | null;
  closing_time: string | null;
  booking_required: boolean;
  booking_link: string | null;
}

export interface StopActivity {
  id: number;
  stop_id: number;
  activity_id: number;
  activity: Activity;
  order_index: number;
  scheduled_date: string | null;
  scheduled_time: string | null;
  notes: string | null;
  is_completed: boolean;
  cost_override_usd: number | null;
}

export interface StopActivityCreateInput {
  activity_id: number;
  scheduled_date?: string;
  scheduled_time?: string;
  notes?: string;
  cost_override_usd?: number;
}

File: backend/app/models/activity.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import (
    Boolean,
    Column,
    Enum,
    ForeignKey,
    Index,
    Numeric,
    String,
    Text,
    Time,
)
from sqlalchemy import DateTime, Uuid as UUID

from app.extensions import db
from app.models.enums import ActivityCategoryEnum


class Activity(db.Model):
    __tablename__ = "activities"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    city_id = Column(
        UUID(as_uuid=True),
        ForeignKey("cities.id", ondelete="CASCADE"),
        nullable=True,
    )
    name = Column(String(200), nullable=False)
    description = Column(Text, nullable=True)
    category = Column(
        Enum(ActivityCategoryEnum, name="activity_category_enum"), nullable=False
    )
    cost_usd = Column(Numeric(8, 2), nullable=False, default=0)
    duration_hours = Column(Numeric(4, 1), nullable=True)
    image_url = Column(Text, nullable=True)
    map_link = Column(Text, nullable=True)
    opening_time = Column(Time, nullable=True)
    closing_time = Column(Time, nullable=True)
    booking_required = Column(Boolean, nullable=False, default=False)
    booking_link = Column(Text, nullable=True)
    created_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )

    city = db.relationship("City", back_populates="activities")
    stop_activities = db.relationship("StopActivity", back_populates="activity", lazy="dynamic")

    __table_args__ = (
        Index("activities_city_id_idx", "city_id"),
        Index("activities_category_idx", "category"),
    )

    def __repr__(self) -> str:
        return f"<Activity {self.name}>"
